contracts/meta-asset-token/DllrTransferWithPermit.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import "./MetaAssetToken.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/draft-ERC20PermitUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/ERC1967/ERC1967UpgradeUpgradeable.sol";
+
+interface IERC20PermitWithTransfer {
+    /**
+     * @dev Sets `value` as the allowance of `spender` over ``owner``'s tokens,
+     * given ``owner``'s signed approval.
+     *
+     * IMPORTANT: The same issues {IERC20-approve} has related to transaction
+     * ordering also apply here.
+     *
+     * Emits an {Approval} event.
+     *
+     * Requirements:
+     *
+     * - `spender` cannot be the zero address.
+     * - `deadline` must be a timestamp in the future.
+     * - `v`, `r` and `s` must be a valid `secp256k1` signature from `owner`
+     * over the EIP712-formatted function arguments.
+     * - the signature must use ``owner``'s current nonce (see {nonces}).
+     *
+     * For more information on the signature format, see the
+     * https://eips.ethereum.org/EIPS/eip-2612#specification[relevant EIP
+     * section].
+     */
+    function permit(
+        address owner,
+        address spender,
+        uint256 value,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external;
+
+    /**
+     * @notice Only owner who can transfer the token.
+     * @notice destination cannot be:
+     * - zero (0x0) address.
+     *
+     * @param _from Sender of the token.
+     * @param _to Recipient of the token.
+     * @param _amount The amount of token that will be transferred.
+     *
+     * @return true / false.
+     */
+    function transferFrom(address _from, address _to, uint256 _amount) external returns (bool);
+
+    /**
+     * @dev Returns the remaining number of tokens that `spender` will be
+     * allowed to spend on behalf of `owner` through {transferFrom}. This is
+     * zero by default.
+     *
+     * This value changes when {approve} or {transferFrom} are called.
+     */
+    function allowance(address owner, address spender) external view returns (uint256);
+}
+
+/**
+ * @dev This is an intermediary contract to fix the griefing attack vulnerbility in the DLLR contract.
+ */
+contract DllrTransferWithPermit is
+    ERC20PermitUpgradeable,
+    OwnableUpgradeable,
+    ERC1967UpgradeUpgradeable
+{
+    /** MODIFIER */
+    modifier requireValidRecipient(address _recipient) {
+        require(
+            _recipient != address(0) && _recipient != address(this),
+            "DLLR: Invalid address. Cannot transfer DLLR to the null address."
+        );
+        _;
+    }
+
+    /** EVENT */
+    /**
+     * @dev Emitted when transfer  Manager config is changed.
+     */
+    event TransferWithPermit(address _from, address _to, uint256 _amount);
+
+    /** STORAGE */
+    IERC20PermitWithTransfer public dllr;
```

Comment:
I don't think we can use immutable here since this is the upgradeable contract, and the dllr is only set from the initializer contract, not from he constructor

Reply:
we can but then it should be initialized in constructor because immutables are stored as a part of the bytecode, not variables, just like constants: https://soliditylang.org/blog/2020/05/13/immutable-keyword/

Reply:
updated in this [commit](https://github.com/DistributedCollective/mynt/pull/17/commits/0b495d7b734a64db865d48d07a012ef1cefba6b4)